test/puppetlabs/trapperkeeper/services_test.clj

```diff
@@ -49,56 +51,137 @@
+        services
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq)))
       (is (= [:init-service1 :init-service2 :init-service3
-              :start-service1 :start-service2 :start-service3]
+              :start-service1 :start-service2 :start-service3
+              :stop-service3 :stop-service2 :stop-service1]
              @call-seq))
+
+
       (reset! call-seq [])
-      (bootstrap-services-with-empty-config [service3 service2 service1])
+      (with-app-with-empty-config app
+        services
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq)))
       (is (= [:init-service1 :init-service2 :init-service3
-              :start-service1 :start-service2 :start-service3]
-             @call-seq))))
+              :start-service1 :start-service2 :start-service3
+              :stop-service3 :stop-service2 :stop-service1]
+             @call-seq)))))
 
+(deftest test-lifecycle-function-ordering-restart
+  (testing "app restart calls life cycle functions in the correct order"
+    (let [call-seq (atom [])
+          services (create-lifecycle-services call-seq)]
+      (with-app-with-empty-config app
+        services
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq))
+        (app/restart app)
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3
+                :stop-service3 :stop-service2 :stop-service1
+                :init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq)))
+      (is (= [:init-service1 :init-service2 :init-service3
+              :start-service1 :start-service2 :start-service3
+              :stop-service3 :stop-service2 :stop-service1
+              :init-service1 :init-service2 :init-service3
+              :start-service1 :start-service2 :start-service3
+              :stop-service3 :stop-service2 :stop-service1]
+             @call-seq)))))
+
+(deftest test-lifecycle-function-ordering-signaling
+  (testing "app restart calls life cycle functions in the correct order"
+    (let [call-seq (atom [])
+          services (create-lifecycle-services call-seq)]
+      (with-app-with-empty-config app
+        services
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq))
+        (internal/register-sighup-handler [app])
+        (beckon/raise! "HUP")
+        (let [start (System/currentTimeMillis)]
+          (while (and (not= (count @call-seq) 15)
+                     (< (- (System/currentTimeMillis) start) 1000))
+            (Thread/yield)))
```

Comment:
I don't think we need this for the signal handler, but it seemed slightly better than a sleep.